File: dotnet/src/ProcessEngine.ExternalTaskAPI.Client/ExternalTaskWorker.cs
namespace ProcessEngine.ExternalTaskAPI.Client
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using System.Timers;
    using Foundation.IAM.Contracts;
    using ProcessEngine.ExternalTaskAPI.Contracts;

    /// <summary>
    /// Periodically fetches, locks and processes ExternalTasks for a given topic.
    /// </summary>
    public class ExternalTaskWorker : IExternalTaskWorker
    {
        private const int LockDurationInMilliseconds = 30000;
        private readonly IExternalTaskAPI externalTaskAPI;

        /// <summary>
        /// Creates an new instance of ExternalTaskWorker
        /// </summary>
        public ExternalTaskWorker(IExternalTaskAPI externalTaskAPI)
        {
            this.externalTaskAPI = externalTaskAPI;
        }

        /// <summary>
        /// Id of worker
        /// </summary>
        public string WorkerId { get; } = Guid.NewGuid().ToString();

        /// <summary>
        /// Periodically fetches, locks and processes available ExternalTasks with a given topic,
        /// using the given callback as a processing function.
        /// </summary>
        /// <param name="identity">
        /// The identity to use for fetching and processing ExternalTasks.
        /// </param>
        /// <param name="topic">
        /// The topic by which to look for and process ExternalTasks.
        /// </param>
        /// <param name="maxTasks">
        /// max. ExternalTasks to fetch.
        /// </param>
        /// <param name="longpollingTimeout">
        /// Longpolling Timeout in ms.
        /// </param>
        /// <param name="handleAction">
        /// The function for processing the ExternalTasks.
        /// </param>
        public async Task WaitForHandle<TPayload>(
            IIdentity identity,
            string topic,
            int maxTasks,
            int longpollingTimeout,
            HandleExternalTaskAction<TPayload> handleAction
        )
            where TPayload : new()
        {
            while (true)
            {
                var externalTasks = await this.FetchAndLockExternalTasks<TPayload>(identity, topic, maxTasks, longpollingTimeout);

                Timer timer = this.StartExtendLockTimer(identity, externalTasks, LockDurationInMilliseconds - 5000);

                IList<Task> tasks = new List<Task>();

                foreach (ExternalTask<TPayload> externalTask in externalTasks)
                {
                    tasks.Add(this.ExecuteExternalTask<TPayload>(identity, externalTask, handleAction));
                }

                await Task.WhenAll(tasks);

                timer.Stop();
            }
        }

        private Timer StartExtendLockTimer<TPayload>(IIdentity identity, IEnumerable<ExternalTask<TPayload>> externalTasks, int intervall)
        {
            Timer timer = new Timer(intervall);
            timer.Elapsed += async (sender, e) => await ExtendLocks<TPayload>(identity, externalTasks.ToList());
            timer.Start();

            return timer;
        }

        private async Task<IEnumerable<ExternalTask<TPayload>>> FetchAndLockExternalTasks<TPayload>(
            IIdentity identity,
            string topic,
            int maxTasks,
            int longpollingTimeout
        )
        where TPayload : new()
        {
            try
            {
                return await this.externalTaskAPI.FetchAndLockExternalTasks<TPayload>(
                  identity,
                  WorkerId,
                  topic,
                  maxTasks,
                  longpollingTimeout,
                  LockDurationInMilliseconds
                );
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);
                await Task.Delay(1000);
                return await this.FetchAndLockExternalTasks<TPayload>(identity, topic, maxTasks, longpollingTimeout);
            }
        }

        private async Task ExecuteExternalTask<TPayload>(
          IIdentity identity,
          ExternalTask<TPayload> externalTask,
          HandleExternalTaskAction<TPayload> handleAction
        )
          where TPayload : new()
        {
            try
            {
                IExternalTaskResult result = await handleAction(externalTask);
                await result.SendToExternalTaskApi(this.externalTaskAPI, identity, WorkerId);

            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);
                await this.externalTaskAPI.HandleServiceError(identity, WorkerId, externalTask.Id, exception.Message, exception.StackTrace);
            }
        }

        private async Task ExtendLocks<TPayload>(IIdentity identity, IList<ExternalTask<TPayload>> externalTasks)
        {
            foreach (ExternalTask<TPayload> externalTask in externalTasks)
            {
                await this.externalTaskAPI.ExtendLock(identity, WorkerId, externalTask.Id, LockDurationInMilliseconds);
            }
        }
    }
}

File: python/src/external_task_worker.py
import asyncio
import threading
import uuid


class ExternalTaskWorker:

    def __init__(self, externalTaskApi):
        self.__lockDuration = 30000
        self.__externalTaskApi = externalTaskApi
        self.workerId = str(uuid.uuid4())

    async def waitForHandle(self, identity, topic, maxTasks, longPollingTimeout, handleAction):
        while True:
            externalTasks = await self.__fetchAndLockExternalTasks(identity, topic, maxTasks, longPollingTimeout)

            timer = self.__startExtendLockTimer(
                identity, externalTasks, (self.__lockDuration - 5000) / 1000)

            try:
                tasks = []

                for externalTask in externalTasks:
                    tasks.append(self.__executeExternalTask(
                        identity, externalTask, handleAction))

                if len(tasks) > 0:
                    await asyncio.wait(tasks)
            finally:
                timer.cancel()

    async def __fetchAndLockExternalTasks(self, identity, topicName, maxTasks, longPollingTimeout):
        try:
            return await self.__externalTaskApi.fetchAndLockExternalTasks(
                identity, self.workerId, topicName, maxTasks, longPollingTimeout, self.__lockDuration)
        except Exception as exception:
            print(exception)

            await asyncio.sleep(1)
            return await self.__fetchAndLockExternalTasks(
                identity, topicName, maxTasks, longPollingTimeout)

    def __extendLocks(self, identity, externalTasks):
        for externalTask in externalTasks:
            asyncio.run(self.__externalTaskApi.extendLock(
                identity, self.workerId, externalTask["id"], self.__lockDuration))

    def __startExtendLockTimer(self, identity, externalTasks, interval):
        timer = threading.Timer(
            interval, self.__extendLocks, args=[identity, externalTasks])
        timer.start()

        return timer

    async def __executeExternalTask(self, identity, externalTask, handleAction):
        try:
            result = await handleAction(externalTask)

            await result.sendToExternalTaskApi(
                self.__externalTaskApi, identity, self.workerId)
        except Exception as exception:
            print(exception)
            await self.__externalTaskApi.handleServiceError(identity, self.workerId, externalTask["id"], exception, "")
